pkg/cli/set/data.go

```diff
@@ -0,0 +1,443 @@
+			}
+		}
+		return valid && changed, err
+	})
+
+	for _, patch := range patches {
+		info := patch.Info
+		name := getObjectName(info)
+		if patch.Err != nil {
+			allErrs = append(allErrs, fmt.Errorf("error: %s %v\n", name, patch.Err))
+			continue
+		}
+
+		if string(patch.Patch) == "{}" || len(patch.Patch) == 0 {
+			klog.V(1).Infof("info: %s was not changed\n", name)
+			continue
+		}
+
+		if o.Local || o.DryRun {
+			if err := o.Printer.PrintObj(info.Object, o.Out); err != nil {
+				allErrs = append(allErrs, err)
+			}
+			continue
+		}
+
+		actual, err := o.Client.Resource(info.Mapping.Resource).Namespace(info.Namespace).Patch(info.Name, types.StrategicMergePatchType, patch.Patch, metav1.PatchOptions{})
+		if err != nil {
+			allErrs = append(allErrs, err)
+			continue
+		}
+
+		if err := o.Printer.PrintObj(actual, o.Out); err != nil {
+			allErrs = append(allErrs, err)
+		}
+	}
+	return utilerrors.NewAggregate(allErrs)
+}
+
+func isBinary(data []byte) bool {
+	for _, b := range data {
+		if b > 127 || (b < 32 && b != 9 && b != 10 && b != 13) {
+			return true
+		}
+	}
+	return false
+}
+
+func updateDataForObject(obj runtime.Object, fn func(data map[string][]byte) error) (bool, error) {
+	switch t := obj.(type) {
+	case *corev1.Secret:
+		if t.Data == nil {
+			t.Data = make(map[string][]byte)
+		}
+		return true, fn(t.Data)
+		// ReplicationController
+	case *corev1.ConfigMap:
+		if t.BinaryData == nil {
+			t.BinaryData = make(map[string][]byte)
+		}
+		for k, v := range t.Data {
+			t.BinaryData[k] = []byte(v)
+		}
+		t.Data = nil
+		if err := fn(t.BinaryData); err != nil {
+			return true, err
+		}
+		for k, v := range t.BinaryData {
+			if isBinary(v) {
+				continue
+			}
+			delete(t.BinaryData, k)
+			if t.Data == nil {
+				t.Data = make(map[string]string)
+			}
+			t.Data[k] = string(v)
+		}
+		return true, nil
+
+		// Deployment
+	default:
+		return false, fmt.Errorf("the object is not a config map or secret and cannot have data updated: %T", t)
+	}
+}
+
+func keyValuesFromFileSources(fileSources []string) (map[string]string, error) {
+	data := make(map[string]string)
+	for _, fileSource := range fileSources {
+		keyName, filePath, err := util.ParseFileSource(fileSource)
+		if err != nil {
+			return nil, err
+		}
+		info, err := os.Stat(filePath)
+		if err != nil {
+			switch err := err.(type) {
+			case *os.PathError:
+				return nil, fmt.Errorf("error reading %s: %v", filePath, err.Err)
+			default:
+				return nil, fmt.Errorf("error reading %s: %v", filePath, err)
+			}
+		}
+		if info.IsDir() {
+			if strings.Contains(fileSource, "=") {
+				return nil, fmt.Errorf("cannot give a key name for a directory path")
+			}
+			fileList, err := ioutil.ReadDir(filePath)
+			if err != nil {
+				return nil, fmt.Errorf("error listing files in %s: %v", filePath, err)
+			}
+			for _, item := range fileList {
+				itemPath := filepath.Join(filePath, item.Name())
+				if item.Mode().IsRegular() {
```

Comment:
\<trollmode\>Oh, that's nice, I wonder when someone will start asking for link support :wink:\</trollmode\>